IBX-5706: DDEV usage (#1983)
ibexa/documentation-developer: docs/getting_started/install_with_ddev.md

## Installation

### 1. Create a DDEV project directory

Start by creating a new directory for your DDEV project by using the following command, where `<my-ddev-project>` stands for your desired directory name:

```bash
mkdir my-ddev-project && cd my-ddev-project
```

### 2. Configure DDEV

#### Configure PHP version and document root

Next, configure your DDEV environment with the following command:

```bash
ddev config --project-type=php --php-version 8.1 --docroot=public --create-docroot
```

This command sets the project type to PHP, the PHP version to 8.1, the document root to `public` directory, and creates the document root.

#### Use another database type (optional)

By default, DDEV uses MariaDB.

To use PostgreSQL instead, run the following command:

```bash
ddev config --database=postgres:15
```

To use MySQL instead, run the following command:

```bash
ddev config --database=mysql:8.0
```

You can also use other versions of MariaDB, Mysql or PostgreSQL. 
See [DDEV database types documentation](https://ddev.readthedocs.io/en/latest/users/extend/database-types/) for available version ranges.

#### Configure database connection

> Now, configure the database connection for your Ibexa DXP project. Depending on your database of choice (MySQL or PostgreSQL), use the appropriate command below.

Reviewer: ```suggestion
Now, configure the database connection for your [[= product_name =]] project. Depending on your database of choice (MySQL or PostgreSQL), use the appropriate command below.
```

Proceed by starting your DDEV environment with the following command:

```bash
ddev start
```

!!! tip

    If you forgot some part of configuration, you can set it by using `ddev config` later, but afterwards you have to restart DDEV with `ddev restart`.

### 4. Composer authentication

Next, you need to [set up authentication tokens](install_ibexa_dxp.md#set-up-authentication-tokens) by modifying the Composer configuration.
You must run the following command **after** executing `ddev start`, because the command runs inside the container.
Replace `<installation-key>` and `<token-password>` with your actual installation key and token password, respectively.

```bash
ddev composer config --global http-basic.updates.ibexa.co <installation-key> <token-password>
```

This authentication won't persist if the project is restarted (by `ddev restart` or `ddev composer create`). You can back up the authentication file (`auth.json`) by using the following command:

```bash
ddev exec "mkdir -p .ddev/homeadditions/.composer; cp ~/.composer/auth.json .ddev/homeadditions/.composer"
```

If you want to reuse an existing `auth.json` file, see [Using an auth.json](#using-an-authjson).

> ### 5. Create Ibexa DXP project

Reviewer: ```suggestion
### 5. Create project
```

If possible, try to avoid using the product name in headers and titles to limit the amount of entries that will have to be added to plugins.yml if we ever decide to change the product name.

> Once DDEV is running, use Composer to create a new Ibexa DXP project. Remember to replace `<edition>` and `<version>` with your desired edition and version respectively.

Reviewer: ```suggestion
Once DDEV is running, use Composer to create a new [[= product_name =]] project. 
Remember to replace `<edition>` and `<version>` with your desired edition and version respectively.
```

```bash
ddev composer create ibexa/<edition>-skeleton:<version>
```

!!! tip

    Consider adding the Symfony Debug bundle which fixes memory outage when dumping objects with circular references.
    `ddev composer require --dev symfony/debug-bundle`

> ### 6. Install the DXP and its database

Reviewer: ```suggestion
### 6. Install the platform and its database
```

> Once you've made this change, you can proceed to install Ibexa DXP.

Reviewer: ```suggestion
Once you've made this change, you can proceed to install [[= product_name =]].
```

```bash
ddev php bin/console ibexa:install ibexa-<edition>
ddev php bin/console ibexa:graphql:generate-schema
```

### 7. Open browser

> Once the above steps are completed, open your Ibexa DXP webpage by running the `ddev launch` command.

Reviewer: ```suggestion
Once the above steps are completed, open the [[= product_name =]]'s webpage by running the `ddev launch` command.
```

!!! tip

    You can also see the project URL in the `ddev start` output.

> ### 8. Start using Ibexa DXP

Reviewer: ```suggestion
### 8. Start using [[= product_name =]]
```

> You can now start using Ibexa DXP and implement your own website on the platform.

Reviewer: ```suggestion
You can now start using [[= product_name =]] and implement your own website on the platform.
```

### Using `auth.json`

An `auth.json` file can be used for one project, or globally for all projects, with the [DDEV `homeaddition` feature](https://ddev.readthedocs.io/en/latest/users/extend/in-container-configuration/).

For example, you can copy an `auth.json` file to a DDEV project:
`cp <path-to-an>/auth.json .ddev/homeadditions/.composer`

Alternatively, the Composer global `auth.json` can be the DDEV global `auth.json` with the help of a symbolic link:
`mkdir -p ~/.ddev/homeadditions/.composer && ln -s ~/.composer/auth.json ~/.ddev/homeadditions/.composer/auth.json`

If the DDEV project has already been started, you need to run `ddev restart`.

The use of an `auth.json` file replaces step [4. Composer authentication](#4-composer-authentication).

### Using Dotenv

Instead of using environment variables inside the container, a [`.env.local`](https://symfony.com/doc/5.4/configuration.html#overriding-environment-values-via-env-local) file can be added to the project.

The following example shows the use of `.env.local` with database configuration:

- Skip step [2. Configure DDEV / Configure database connection](#configure-database-connection).

> - Modify step [5. Create Ibexa DXP project](#5-create-ibexa-dxp-project) to insert the database setting:

Reviewer: ```suggestion
- Modify step [5. Create [[= product_name =]] project](#5-create-ibexa-dxp-project) to insert the database setting:
```

  ```bash
  ddev composer create ibexa/commerce-skeleton --no-install;
  echo "DATABASE_URL=mysql://db:db@db:3306/db" >> .env.local;
  ddev composer install;
  ```

!!! note "Precedence"

    For the same variable, its server level environment value overrides its application level `.env` value.
    To switch a variable from `ddev config --web-environment-add` command to `.env.local` file, you have to do either of the following:

    - remove it from under the `web_environment:` key in `.ddev/config.yaml` file then restart the project
    - rebuild the project from scratch

### Nginx Server Blocks

> Even if Ibexa DXP works with the default Nginx configuration coming with DDEV, it's recommended to use the dedicated one.

Reviewer: ```suggestion
Even if [[= product_name =]] works with the default Nginx configuration that comes with DDEV, it's recommended to use a dedicated one.
```

| `%BASEDIR%`             | `/var/www/html`              |
| `%BODY_SIZE_LIMIT_M%`   | `0`                          |
| `%TIMEOUT_S%`           | `90s`                        |
| `%FASTCGI_PASS%`        | `unix:/var/run/php-fpm.sock` |
| `%BINARY_DATA_HANDLER%` | empty string                 |

Because of path resolution inside DDEV's Nginx, you must replace one more thing: `ibexa_params.d` with `sites-enabled/ibexa_params.d`.

You can, for example, do it with `sed`:

```bash
sed -i 's/%PORT%/80/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%HOST_LIST%/*.ddev.site/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%BASEDIR%/\/var\/www\/html/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%BODY_SIZE_LIMIT_M%/0/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%TIMEOUT_S%/90s/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%FASTCGI_PASS%/unix:\/var\/run\/php-fpm.sock/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%BINARY_DATA_HANDLER%//' .ddev/nginx_full/ibexa.conf;
sed -i 's/ibexa_params.d/sites-enabled\/ibexa_params.d/' .ddev/nginx_full/ibexa.conf;
```

### Switch to Apache and its Virtual Host

To use Apache instead of the default Nginx, run the following command:

```bash
ddev config --webserver-type=apache-fpm
```

> Ibexa DXP can't run on Apache without the dedicated Virtual Host.

Reviewer: ```suggestion
[[= product_name =] can't run on Apache without a dedicated Virtual Host.
```

> To set the Apache Virtual Host, override `.ddev/apache/apache-site.conf` with Ibexa DXP's config. You can do it manually or by using a script.

Reviewer: ```suggestion
To set the Apache Virtual Host, override `.ddev/apache/apache-site.conf` with [[= product_name =]]'s config. 
You can do it manually or by using a script.
```

> #### Manual

Reviewer: ```suggestion
#### Manual procedure
```

ddev start
# Configure Composer authentication
ddev composer config --global http-basic.updates.ibexa.co <installation-key> <token-password>
# Install the dependencies packages
ddev composer install
# Populate the database with a clean install
ddev php bin/console ibexa:install ibexa-<edition>
# Add some content types using a migration file (previously created on another installation) and update the GraphQL schema
ddev php bin/console ibexa:migrations:migrate --file=project_content_types.yaml
ddev php bin/console ibexa:graphql:generate-schema
# Open the project in the default brother which should display the default SiteAccess frontpage
ddev launch
```

Notice that the example adds the whole `.ddev/` directory to `.gitignore`, but you can also version parts of it.
Some DDEV configs can be shared among developers. For example, a common `.ddev/config.yaml` can be committed for everyone and [locally extended or overridden](https://ddev.readthedocs.io/en/latest/users/extend/customization-extendibility/#extending-configyaml-with-custom-configyaml-files).

Compared to running a clean install like described in [Installation steps](#installation-steps), you can proceed as follows:

> - In [1. Create a DDEV project directory](#1-create-a-ddev-project-directory), you can use an existing directory that contains an Ibexa DXP project instead of creating an empty directory.

Reviewer: ```suggestion
- In [1. Create a DDEV project directory](#1-create-a-ddev-project-directory), you can use an existing directory that contains an [[= product_name =]] project instead of creating an empty directory.
```

- In [2. Configure DDEV / Configure PHP version and document root](#configure-php-version-and-document-root), don't create the Document root, remove the `--create-docroot` option.

> - In [5. Create Ibexa DXP project](#5-create-ibexa-dxp-project), use only `ddev composer install` instead of `ddev composer create`.

Reviewer: ```suggestion
- In [5. Create [[= product_name =]] project](#5-create-ibexa-dxp-project), use only `ddev composer install` instead of `ddev composer create`.
```

- Populate the database with [Ibexa data migration](../content_management/data_migration/importing_data.md) or [`ddev import-db`](https://ddev.readthedocs.io/en/latest/users/usage/commands/#import-db).

### Hostnames and domains

If the local project needs to answer to real production domains (for example, to use the existing [hostname to SiteAccess](../multisite/siteaccess/siteaccess_matching.md#maphost) or [hostname element to SiteAccess](../multisite/siteaccess/siteaccess_matching.md#hostelement) mappings), you can use [additional hostnames](https://ddev.readthedocs.io/en/latest/users/extend/additional-hostnames/).

!!! caution

    As this feature modifies domain resolution, the real website may be unreachable until the `hosts` file is manually cleaned.

### Cluster or Ibexa Cloud

DDEV can be useful to locally simulate a production cluster.

- See [clustering with DDEV](clustering_with_ddev.md) to add Elasticsearch, Solr, Redis or Memcached to your DDEV installation.

> - See [Ibexa Cloud and DDEV](ibexa_cloud_and_ddev.md) to locally run an Ibexa DXP project by using DDEV.

Reviewer: ```suggestion
- See [Ibexa Cloud and DDEV](ibexa_cloud_and_ddev.md) to locally run an [[= product_name =]] project by using DDEV.
```

| `%BASEDIR%`             | `/var/www/html`              |
| `%BODY_SIZE_LIMIT_M%`   | `0`                          |
| `%TIMEOUT_S%`           | `90s`                        |
| `%FASTCGI_PASS%`        | `unix:/var/run/php-fpm.sock` |
| `%BINARY_DATA_HANDLER%` | empty string                 |

Because of path resolution inside DDEV's Nginx, you must replace one more thing: `ibexa_params.d` with `sites-enabled/ibexa_params.d`.

You can, for example, do it with `sed`:

```bash
sed -i 's/%PORT%/80/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%HOST_LIST%/*.ddev.site/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%BASEDIR%/\/var\/www\/html/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%BODY_SIZE_LIMIT_M%/0/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%TIMEOUT_S%/90s/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%FASTCGI_PASS%/unix:\/var\/run\/php-fpm.sock/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%BINARY_DATA_HANDLER%//' .ddev/nginx_full/ibexa.conf;
sed -i 's/ibexa_params.d/sites-enabled\/ibexa_params.d/' .ddev/nginx_full/ibexa.conf;
```

### Switch to Apache and its Virtual Host

To use Apache instead of the default Nginx, run the following command:

```bash
ddev config --webserver-type=apache-fpm
```

> [[= product_name =] can't run on Apache without a dedicated Virtual Host.

Author: ```suggestion
[[= product_name =]] can't run on Apache without a dedicated Virtual Host.
```

  --host-alias='*.ddev.site' \
  --basedir='/var/www/html' \
  --sf-env=dev \
  > .ddev/apache/apache-site.conf
sed -i 's/php5-fpm.sock/php-fpm.sock/' .ddev/apache/apache-site.conf
rm vhost.sh
```

Then, restart the project:

```bash
ddev restart
```

### Run an already existing project

To run an existing project, you need to:

1. Configure the DDEV project.
1. Start the DDEV project.
1. Add Composer authentication.
1. Install dependencies packages with Composer.
1. Populate the contents, which could mean:
    - getting a clean database with ddev `php bin/console ibexa:install ibexa-<edition>` and adding some data with [Ibexa data migration](../content_management/data_migration/importing_data.md), or
    - injecting a dump with [`ddev import-db`](https://ddev.readthedocs.io/en/latest/users/usage/commands/#import-db) and copying related binary files into `public/var`.

The following examples run an already [version-controlled project](install_ibexa_dxp.md#add-project-to-version-control) and have the right content structure (but no content):

```bash
# Clone the version-controlled project and enter its local directory
git clone <repository> my-ddev-project && cd my-ddev-project

> # Exclude the whole `.ddev/` directory from version control

Author: @mnocon Can I go with something as simple as the following or is it needing more highlight?

```suggestion
# Exclude the whole `.ddev/` directory from version control (some DDEV config could have been committed and shared, see notice below)
```

    - run [`ddev help <command>`](https://ddev.readthedocs.io/en/latest/users/usage/commands/#help) to get usage details about a specific command
    
    Learn more about DDEV configuration from [`ddev config` command documentation](https://ddev.readthedocs.io/en/latest/users/usage/commands/#config) and [advanced configuration files documentation](https://ddev.readthedocs.io/en/latest/users/configuration/config/).

### Using `auth.json`

An `auth.json` file can be used for one project, or globally for all projects, with the [DDEV `homeaddition` feature](https://ddev.readthedocs.io/en/latest/users/extend/in-container-configuration/).

For example, you can copy an `auth.json` file to a DDEV project:
`cp <path-to-an>/auth.json .ddev/homeadditions/.composer`

Alternatively, the Composer global `auth.json` can be the DDEV global `auth.json` with the help of a symbolic link:
`mkdir -p ~/.ddev/homeadditions/.composer && ln -s ~/.composer/auth.json ~/.ddev/homeadditions/.composer/auth.json`

If the DDEV project has already been started, you need to run `ddev restart`.

> The use an `auth.json` file replaces step [4. Composer authentication](#4-composer-authentication).

Author: ```suggestion
The use of an `auth.json` file replaces step [4. Composer authentication](#4-composer-authentication).
```
